labs/cryptol-overview/overview.md

```diff
@@ -0,0 +1,544 @@
+
+## Primitives
+
+Cryptol offers a robust set of primitives to construct more complex functions and specifications with. The built-in primitives cover a wide variety of purposes. A sampling of these built-ins include:
+
+ * **List manipulation** -- `take`, `drop`
+ 
+```haskell
+Main> take `{1} [1, 2, 3]
+[1]
+Main> drop `{1} [1, 2, 3]
+[2, 3]
+```
+ 
+ * **Structural manipulation** -- `split`, `groupBy`
+
+```haskell
+Main> split `{3} [1, 2, 3, 4, 5, 6]
+[[1, 2], [3, 4], [5, 6]]
+Main> groupBy `{3} [1, 2, 3, 4, 5, 6]
+[[1, 2, 3], [4, 5, 6]]
+```
+
+ * **Arithmetic** -- `sum`, `min`, `max`
+
+```haskell
+Main> sum [1, 2, 3, 4, 5]
+15
+Main> min 5 10
+5
+Main> max 5 10
+10
+```
+
+Inside of the Cryptol interpreter you can inspect the symbols, functions, and primitives that are defined within the current context using the `:browse` (or simply `:b`) command. Doing so at the beginning of a fresh interpreter session will show you the list of built-ins and other information about the environment:
+
+```haskell
+Main> :b
+    all : {n, a} (fin n) => (a -> Bit) -> [n]a -> Bit
+    and : {n} (fin n) => [n] -> Bit
+    any : {n, a} (fin n) => (a -> Bit) -> [n]a -> Bit
+    ...
+    zext : {m, n} (fin m, m >= n) => [n] -> [m]
+    zip : {n, a, b} [n]a -> [n]b -> [n](a, b)
+    zipWith : {n, a, b, c} (a -> b -> c) -> [n]a -> [n]b -> [n]c
+```
+
+
+## Functions
+
+Cryptol allows users to define functions allowing subcomponents to be reused or allow greater clarity in a program. Functions are typically specified in two parts: the *type definition* and the *function definition*.
+
+Suppose that a project file `rotword.cry` contains the following code:
+
+```haskell
+RotWord : [4][8] -> [4][8]
+RotWord [a0, a1, a2, a3] = [a1, a2, a3, a0]
+```
+
+This provides a type for `RotWord` (`[4][8] -> [4][8]`) and a definition (`RotWord [a0, a1, a2, a3] = [a1, a2, a3, a0]`). Within the interpreter we can perform computations and check the type of this function as follows
+
+```haskell
+Main> RotWord [0x01, 0x02, 0x03, 0x04]
+[0x02, 0x03, 0x04, 0x01]
+Main> :t RotWord 
+RotWord : [4][8] -> [4][8]
+```
+
+To enjoy greater overall confidence in a system we are developing, we may want to check properties about the components we build. Suppose we wanted to check that `RotWord` was the identity on sequences where all the elements were the same. We could add the following line to `rotword.cry`:
+
+```haskell
+RotWord : [4][8] -> [4][8]
+RotWord [a0, a1, a2, a3] = [a1, a2, a3, a0]
+
+property check_identity x = RotWord [x, x, x, x] == [x, x, x, x]
+```
+
+and then from the interpreter we could check that this property was true:
+
+```haskell
+Cryptol> :l rotword.cry 
+Loading module Cryptol
+Loading module Main
+Main> :prove
+:prove check_identity
+	Q.E.D.
+(Total Elapsed Time: 0.005s, using Z3)
+```
+
+`Q.E.D.` Building up more complex properties and relationships between the components of a system allow us to assert with very high confidence that the system is correct.
```

Comment:
Qed seems out of place here

Reply:
Removed, that was a pause point for me at one point.